### PR TITLE
[profiler] Introduce the Mono.Profiler.Log library. API still unstable.

### PR DESCRIPTION
**Pull request by alexrp**
This is a super early alpha version of the `Mono.Profiler.Log` library. It can read all of the MLPD files I've thrown at it. It should be mergeable as-is, but is missing a few things:

* No runtime profiler interface (requires significant changes to the runtime still).
* No reading of coverage events.
* No support for starting new processes with profiling enabled (as in debugger-libs).
* Event object model will go through some changes as some of my unmerged PRs get updated and merged.

Usage:

```bash
alexrp@alexrp:~/tests$ cat read.cs
using System;
using System.IO;
using Mono.Profiler.Log;

class MyVisitor : LogEventVisitor {

        public override void Visit (AssemblyLoadLogEvent ev)
        {
                Console.WriteLine ("Loaded assembly {0}", ev.Name);
        }

        // etc ...
}

class Program {

        static void Main ()
        {
                var reader = new LogReader (File.OpenRead ("output.mlpd"), false);
                var visitor = new MyVisitor ();
                var proc = new LogProcessor (reader, visitor);

                try {
                        proc.Process ();
                } catch (EndOfStreamException) {
                        // Finished processing the file.
                } catch (LogException ex) {
                        Console.WriteLine (ex);
                }
        }
}
alexrp@alexrp:~/tests$ mcs -r:Mono.Profiler.Log read.cs
alexrp@alexrp:~/tests$ mono read.exe
Loaded assembly mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
Loaded assembly test, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null
alexrp@alexrp:~/tests$
```

I do realize that using `EndOfStreamException` to signal completion is...odd. The problem is that, as far as I can tell, there's no easy way to check if a .NET `Stream` has more data without also advancing the stream in the process. Checking the `Position` against `Length` isn't valid for all stream types. I'm open to ideas here. Maybe just catch the exception internally and return normally.

### CODE REVIEW

File: mcs/class/Makefile [hunk 1]
```diff
@@ -50,6 +50,7 @@ endif
 
 mobile_common_dirs := \
 	corlib	\
+	Mono.Profiler.Log	\
```
Comment (marek-safar) [inline mcs/class/Makefile]: Do we really want to distribute for building mobile apps?

File: mcs/class/Makefile [hunk 2]
```diff
@@ -180,6 +182,7 @@ xammac_4_5_dirs := \
 
 net_4_x_dirs := \
 	corlib				\
+	Mono.Profiler.Log	\
```
Comment (marek-safar) [inline mcs/class/Makefile]: Remove this one, when you have it in parallel list

File: mcs/class/Mono.Profiler.Log/Makefile [hunk 3]
```diff
@@ -0,0 +1,12 @@
+thisdir = class/Mono.Profiler
+include ../../build/rules.make
+
+LIBRARY = Mono.Profiler.Log.dll
+LIBRARY_SNK = ../mono.snk
+
+LIB_REFS = System System.Core
+LIB_MCS_FLAGS = /unsafe /keyfile:$(LIBRARY_SNK) /publicsign
+
+NO_TEST = yes
```
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Makefile]: Please add at least basic tests

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEnums.cs [hunk 4]
```diff
@@ -0,0 +1,198 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+namespace Mono.Profiler.Log {
+
+	enum LogEventType {
```
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEnums.cs]: It should include link to real runtime header file

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Mono.Profiler.Log {
+
+	sealed class LogEventPool {
```
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs]: Is it really worth to have this caching causing more gen1 pressure and causing hard to debug issues when user code still wants to use original reference

**Comment by alexrp** [comment]
@marek-safar @akoeplinger is there anything else I need to do when adding a new assembly like this?

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+					_bufferHeader = new LogBufferHeader (_streamHeader, Reader);
+
+					var goal = Reader.BaseStream.Position + _bufferHeader.Length;
+
+					while (Reader.BaseStream.Position < goal) {
+						token.ThrowIfCancellationRequested ();
+
+						var ev = ReadEvent ();
+
+						events.Add (ev);
```
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: How big the list can be?

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+		{
+			return _bufferHeader.CurrentTime += Reader.ReadULeb128 ();
+		}
+
+		IReadOnlyList<long> ReadBacktrace (bool actuallyRead, bool managed = true)
+		{
+			if (!actuallyRead)
+				return Array.Empty<long> ();
+
+			var list = new List<long> ((int) Reader.ReadULeb128 ());
```
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: Array would be more efficient

**Comment by alexrp** [comment]
What's up with those failures?

```
error CS0006: Metadata file './../../class/lib/net_4_x-linux/System.dll' could not be found
error CS0006: Metadata file './../../class/lib/net_4_x-linux/System.Core.dll' could not be found
make[8]: *** [../../class/lib/net_4_x-linux/Mono.Profiler.Log.dll] Error 1
```

File: mcs/class/Makefile (hunk 1 above)
Comment (alexrp) [inline mcs/class/Makefile]: Yes. This library will, in the future, allow interacting with the profiler while it's running, so having it available on mobile is important.

File: mcs/class/Mono.Profiler.Log/Makefile (hunk 3 above)
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Makefile]: There will be a whole test suite based on this library that will replace the current profiler test suite. That'll be a separate PR.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs (hunk 5 above)
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs]: Yes. The Xamarin Profiler has in fact started doing object pooling exactly like this because the sheer amount of event allocations was proving to be a problem. This design is based around their needs.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs (hunk 6 above)
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: The profiler emits synchronization point events every second. If we go by sampling alone, we generate 100 events per second per thread. So not that big. It's also worth noting that the Xamarin Profiler already does this in its internal MLPD reading library.

Comment (kumpera) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: Tens of thousands events.

File: mcs/class/Makefile (hunk 1 above)
Comment (kumpera) [inline mcs/class/Makefile]: It's ok if we don't ship it anywhere but desktop for now.

Comment (alexrp) [inline mcs/class/Makefile]: I mean, what stops us from shipping it on mobile, exactly?

Comment (akoeplinger) [inline mcs/class/Makefile]: Also you need to move this down to at least below System.Core since it depends on that one. This is the cause of the errors:

```
error CS0006: Metadata file './../../class/lib/monodroid/System.dll' could not be found
error CS0006: Metadata file './../../class/lib/monodroid/System.Core.dll' could not be found
```

In fact, I'd just add it to the end of the list so the more "important" libraries are built first.

File: mcs/class/Makefile [hunk 8]
```diff
@@ -132,6 +133,7 @@ winaot_dirs := \
 
 xammac_4_5_dirs := \
 	corlib	\
+	Mono.Profiler.Log	\
```
Comment (akoeplinger) [inline mcs/class/Makefile]: same here, needs moving.

File: mcs/class/Makefile (hunk 1 above)
Comment (marek-safar) [inline mcs/class/Makefile]: Mobile has limited available memory and this library load everything into memory so it does not seems to me as useful for mobile with current design.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs (hunk 6 above)
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: Then it'd make sense to start with List initial size bigger than 4 elements

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs (hunk 5 above)
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs]: That's fine but don't do that on instances which users work with that's just poor design.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs [hunk 9]
```diff
@@ -0,0 +1,619 @@
...
+				events.Clear ();
+			}
+
+			try {
+				while (true) {
+					token.ThrowIfCancellationRequested ();
+
+					_bufferHeader = new LogBufferHeader (_streamHeader, Reader);
+
+					var goal = Reader.BaseStream.Position + _bufferHeader.Length;
```
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: Should take into account stream.Length when available

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs (hunk 5 above)
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs]: I don't understand what you're proposing here. Those instances are the problem. How would not pooling them solve anything?

File: mcs/class/Makefile (hunk 1 above)
Comment (alexrp) [inline mcs/class/Makefile]: The MLPD reading is not the important part for mobile. The `LogRuntimeProfiler` class, when actually implemented, will contain methods for interacting with the profiler at runtime. The code isn't there yet, but I don't see any harm in including the library for mobile now to make sure it actually builds, especially since changes need to be made in XA/XI for it to be included there anyway, which we can do when the library is ready.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs (hunk 9 above)
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: Since it's only available for some stream types, what good would it actually do? Sure, when it's available, we can detect EOS and return normally, but for stream types where it isn't, we'll still throw an `EndOfStreamException`. So, either way, the caller ends up having to catch that (or we do it internally).

Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: It makes no sense to throw EndOfStreamException on valid input

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs (hunk 5 above)
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs]: I am not proposing anything. I am just commenting on poor API where you are leaking internal optimization into end user code.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs (hunk 9 above)
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: Like I said in the PR description, I'm open to ideas. I don't know any universal way to detect the end of a stream before reading from it with a `BinaryReader`. So one option is that we catch the `EndOfStreamException` when trying to read the buffer header, and just return normally.

Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: Both stream.Length and stream.Position are available on seekable streams only. So not sure what you are trying to say here. For non-seekable streams you probably want different API anyway

Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: That's a good point. The code shouldn't even be using `Position`, then, as it needs to work on e.g. `NetworkStream` for live profiling. That's a simple fix anyway. That still leaves us with the same problem, though.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs (hunk 5 above)
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs]: One thing we can do is have a `Clone` method on all these `LogEvent` classes so that a user can easily make a copy if they really want to keep a reference around.

Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs]: What about changing the logic to use more delay-loading with something like

```csharp
struct MethodCoverageLogEvent
{
    internal long streamPosition;
    MethodCoverageLog ReadData (DataManager dataManager)
    {
        // initialize from stream in dataManager
    }
}

public class MethodCoverageLog
{
    // Record class for method coverage data
}
```

You keep cyclic array of WeakMethodCoverageLogEvent which you reuse after

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs (hunk 9 above)
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: I think for live profiling you should have different API which is based on filters or events

**Comment by vargaz** [comment]
If the log events are small, maybe we could make them a struct ?

**Comment by alexrp** [comment]
@vargaz they can't be a struct because we need to put them in a list and sort them based on timestamp. I can't think of a way to do that without boxing.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs (hunk 9 above)
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: I don't think the profiler team wants to deal with a different API for live vs offline reading. @rodrmoya ?

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs (hunk 6 above)
Comment (rodrmoya) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: In the Xamarin Profiler we use also a List, but was wondering if it made sense to use a better performant collection, like https://github.com/xamarin/profiler/blob/master/src/XamarinProfiler.Core/Framework/LargeList.cs, which avoids the array-resizing-and-instantiation problem List<T> has. We debugged and found SYNC_POINT events come usually after 4/5000 events (of course, depends a lot on the case), so since our LargEList uses 10000 as the size for each array, I wonder if doing something similar  but with smaller, let's say 1000, would bring any benefit.

We haven't profiled that yet, so can't say for sure right now.

**Comment by alexrp** [comment]
I have no idea what's up with the current build failure. Why is Roslyn crashing on every bot?

**Comment by vargaz** [comment]
Since the data is huge, you probably want a preprocessing step which is ran over the raw data without allocating objects for it, and creates a smaller number of higher level objects. I.e. instead of a MethodEntry object, count the number of method entries for the same method, etc.

**Comment by marek-safar** [comment]
That's sort of what I suggested at https://github.com/mono/mono/pull/5019#discussion_r121650194.

You can sort structs in Array with no issue but the question is whether you need to sort at all. I don't know logic behind timestamp but maybe you could do the check on insert to the array.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs (hunk 5 above)
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs]: There are a few issues with this:

1. It won't work with non-seekable streams.
2. How do we sort all events between two sync points by timestamp in this design, while avoiding boxing?
3. You can't actually just read a random event in a log file because interpreting the values in that events requires having read all prior events in that buffer.

Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs]: ad 1, I think that's fine
ad 2, you do at insert
ad 3, you have start point for each event because you read it but not stored it

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs (hunk 9 above)
Comment (marek-safar) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: I believe we must have seekable and non-seekable API versions. There is no way to have performant "rich" API built on top of non-seekable API because we put everything in memory to emulate that.

File: mcs/class/Makefile [hunk 10]
```diff
@@ -1,4 +1,4 @@
-thisdir = class
+hisdir = class
```
Comment (akoeplinger) [inline mcs/class/Makefile]: this is the reason for the current build break :)

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+
+			void ProcessEvents ()
+			{
+				foreach (var ev in events.OrderBy (x => x.Timestamp)) {
+					token.ThrowIfCancellationRequested ();
+
+					ev.Accept (Visitor);
+					_pool.Return (ev);
+				}
+
```
Comment (rodrmoya) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: Another optimisation we recently added to the Xamarin Profiler was about reading the file in one thread and processing it in another (https://github.com/xamarin/profiler/blob/master/src/XamarinProfiler.Core/Platforms/Mono/MlpdReader.cs#L164), so that we can send to the UI live timing information, as well as perf counters, for instance, while we process heavier events (snapshots, etc) on a separate thread. For that, we've had to make the whole API async -> https://github.com/xamarin/profiler/tree/master/src/XamarinProfiler.Core/Platforms/Mono so it would be good if this API had that optimisation in. If not, we'll get back to having the UIs look stuck while processing huge MLPDs

File: mcs/class/Makefile (hunk 10 above)
Comment (alexrp) [inline mcs/class/Makefile]: No idea how that snuck in there...

**Comment by alexrp** [comment]
OK, so reading over the comments, I *think* there may be a misunderstanding of the problem I was trying to solve with object pooling. To be clear: The problem *isn't* that we're getting a ton of irrelevant events that we want to skip over. @kumpera's recent changes to the profiler have largely solved that. Most readers (e.g. the Xamarin Profiler, a managed version of mprof-report, an MLPD linter, etc) will be interested in *all* the events in the file. We're moving to an approach where we disable emitting data we're not interested in rather than discarding/skipping over it in the tools that consume the MLPD. This is really the only reasonable approach for profiling heavy apps like Xamarin Studio.

So, delayed reading doesn't really solve an actual problem that we'll have going forward.

The object pooling was simply meant to alleviate GC pressure caused by rapid allocation of event objects. However, after talking to @rodrmoya, it's become clear that this *probably* isn't as big of a problem as was thought. Most of the GC issues in the Xamarin Profiler came from other areas where they would keep millions of objects around. In reality, most apps only produce a few thousand events per second, which SGen can handle just fine. So I think it's safe to get rid of the object pooling entirely.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs (hunk 5 above)
Comment (kumpera) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEventPool.cs]: @marek-safar The problem is a lot more complicated.

The log *must* be sequentially read, performance tanks otherwise, even on SSDs. Random seeking kill all forms of buffering.

Events must be buffered until we reach a sync point and at that point we sort them. Using value types for the decoded payload would be horribly inefficient as events have an unbound max size or we'd be forced to decode them twice - and the second time would require seeking.

I don't know what you mean by sort at insertion time, there's no inserting time. If you mean at decoding time. That's even more inefficient that sorting at sync points. Remember, we don't report events until we reach a sync point.

Finally, APIs have being using FlyWeight objects since the dawn of time, as long as it's well documented, I think it's fine.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs (hunk 6 above)
Comment (kumpera) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: Segmented arrays have quite slower access times. Why not simply reserve 10k as size and recycle it across decoders?

**Comment by kumpera** [comment]
I see the following issues with the current PR:

- We should ship it only on desktop, but only as long as it's not in a way that it's considered part of our API.

- It should reduce the usage of unsigned types. The ergonomics of them with C# is horrible. You have to add a ton of casts all over the place.

- The exception based control flow must go. It's completely unexpected.
